### PR TITLE
Gnulib update 2023

### PR DESCRIPTION
**Pull request by RincewindsHat**
Update gnulib to stable-202301 branch (668c0b8ffaece623cb0a349e129de4cadd300ff1)

### CODE REVIEW

**Comment by thesamesam** [comment]
It builds on musl but there's an issue which needs fixing:
```
sslutils.c: In function 'np_net_ssl_check_cert':
sslutils.c:198:16: warning: implicit declaration of function 'np_net_ssl_check_certificate'; did you mean 'np_net_ssl_check_cert'? [-Wimplicit-function-declaration]
  198 |         return(np_net_ssl_check_certificate(certificate, days_till_exp_warn, days_till_exp_crit));
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                np_net_ssl_check_cert
  CC       check_swap.o
  CC       check_nt.o
```

We need to be careful with these -- they become fatal in newer compilers (like Clang 16 which is being released soon).

**Comment by RincewindsHat** [comment]
@thesamesam I moved the code a little bit, should be resolved now. (At least that specific bit)

**Comment by thesamesam** [comment]
Looking good!

I got this which is non-fatal with binutils' bfd, but it is fatal with LLVM's lld linker:
```

/bin/sh ../libtool  --tag=CC   --mode=link gcc -DNP_VERSION='"2.3.3.14.g782d6"' -g -O2  -L. -o check_pgsql check_pgsql.o libnpcommon.a ../lib/libmonitoringplug.a ../gl/libgnu.a -lcrypto  -lnsl -lresolv -L/lib -lpq -lcrypt -lpthread -ldl
libtool: link: gcc -DNP_VERSION=\"2.3.3.14.g782d6\" -g -O2 -o check_pgsql check_pgsql.o  -L. libnpcommon.a ../lib/libmonitoringplug.a ../gl/libgnu.a -lcrypto -lnsl -lresolv -L/lib -lpq -lcrypt -lpthread -ldl
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: skipping incompatible /lib/libcrypto.so when searching for -lcrypto
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: skipping incompatible /lib/libnsl.so when searching for -lnsl
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: skipping incompatible /lib/libresolv.so when searching for -lresolv
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: skipping incompatible /lib/libresolv.a when searching for -lresolv
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: skipping incompatible /lib/libcrypt.so when searching for -lcrypt
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: skipping incompatible /lib/libcrypt.a when searching for -lcrypt
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: skipping incompatible /lib/libc.so when searching for -lc
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: skipping incompatible /lib/libc.a when searching for -lc
```

It looks like that leaks in via `PGLIBS` in `configure.ac` (tbh I suspect we could just use PKG_CHECK_MODULES there). But none of this is a regression and I think we could handle all of that separately another time.

On musl, it all builds well too, we get some trivial warnings which are easy to fix (I can fix them another time if I remember):
```
In file included from common.h:123,
                 from netutils.c:30:
/usr/include/sys/poll.h:1:2: warning: #warning redirecting incorrect #include <sys/poll.h> to <poll.h> [-Wcpp]
    1 | #warning redirecting incorrect #include <sys/poll.h> to <poll.h>
      |  ^~~~~~~
In file included from common.h:123,
                 from check_apt.c:36:
/usr/include/sys/poll.h:1:2: warning: #warning redirecting incorrect #include <sys/poll.h> to <poll.h> [-Wcpp]
    1 | #warning redirecting incorrect #include <sys/poll.h> to <poll.h>
      |  ^~~~~~~
In file included from common.h:123,
                 from runcmd.h:27,
                 from runcmd.c:42:
/usr/include/sys/poll.h:1:2: warning: #warning redirecting incorrect #include <sys/poll.h> to <poll.h> [-Wcpp]
    1 | #warning redirecting incorrect #include <sys/poll.h> to <poll.h>
      |  ^~~~~~~
In file included from common.h:123,
                 from check_cluster.c:29:
/usr/include/sys/poll.h:1:2: warning: #warning redirecting incorrect #include <sys/poll.h> to <poll.h> [-Wcpp]
    1 | #warning redirecting incorrect #include <sys/poll.h> to <poll.h>
      |  ^~~~~~~
In file included from common.h:123,
                 from utils.c:25:
/usr/include/sys/poll.h:1:2: warning: #warning redirecting incorrect #include <sys/poll.h> to <poll.h> [-Wcpp]
    1 | #warning redirecting incorrect #include <sys/poll.h> to <poll.h>
      |  ^~~~~~~
In file included from common.h:123,
                 from popen.c:41:
/usr/include/sys/poll.h:1:2: warning: #warning redirecting incorrect #include <sys/poll.h> to <poll.h> [-Wcpp]
    1 | #warning redirecting incorrect #include <sys/poll.h> to <poll.h>
      |  ^~~~~~~
```

So I think we're all good! Thank you for working on this! Would you be able to document the process somewhere (maybe just some rough notes)?

**Comment by RincewindsHat** [comment]
@thesamesam What exactly would you like to have documented?

**Comment by sni** [comment]
image
You're kidding, right? :-)

**Comment by RincewindsHat** [comment]
well, copying the changes of a whole library of ~13 years does indeed change some files :-)

**Comment by sni** [comment]
sure, but the .deps folder should probably not be included in the PR, right?

**Comment by RincewindsHat** [comment]
I don't know, you tell me :-) What does it do?

**Comment by thesamesam** [comment]
> @thesamesam What exactly would you like to have documented?

I wasn't sure how to go about updating the copy locally because `gnulib` didn't seem to be a submodule here and I didn't see a bootstrap script. I was wondering about maybe just a few bullet points about how you got a new copy and integrated it. No bother if not.

**Comment by RincewindsHat** [comment]
@thesamesam It's actually documented https://www.monitoring-plugins.org/doc/faq/gnulib.html (thanks @weiss I suspect).
The rest of the PR is mostly fixing stuff which broke after that (and is hopefully fixed now).

**Comment by thesamesam** [comment]
Ah, thank you! Sorry for missing that!

**Comment by RincewindsHat** [comment]
@sni it's 449 now. That's much better! :-)

File: plugins/check_disk.c [hunk 1]
```diff
@@ -1056,7 +1056,7 @@ get_path_stats (struct parameter_list *p, struct fs_usage *fsp) {
   p->dfree_units = p->available*fsp->fsu_blocksize/mult;
   p->dtotal_units = p->total*fsp->fsu_blocksize/mult;
   /* Free file nodes. Not sure the workaround is required, but in case...*/
-  p->inodes_free  = fsp->fsu_favail > fsp->fsu_ffree ? 0 : fsp->fsu_favail;
```
Comment (sni) [inline plugins/check_disk.c:1059]: is this change required for the gnulib update or something else?

File: plugins/sslutils.c [hunk 2]
```diff
@@ -191,17 +191,6 @@ int np_net_ssl_read(void *buf, int num) {
 	return SSL_read(s, buf, num);
 }
 
-int np_net_ssl_check_cert(int days_till_exp_warn, int days_till_exp_crit){
```
Comment (sni) [inline plugins/sslutils.c:194]: same for this change? Is it part of the gnulib update?

File: plugins/check_disk.c (hunk 1 above)
Comment (RincewindsHat) [reply plugins/check_disk.c:1059]: This one is actually required because `fsu_favail` is not there anymore.

File: plugins/sslutils.c (hunk 2 above)
Comment (RincewindsHat) [reply plugins/sslutils.c:194]: This one is not actually required for the gnulib update, but it fixes a compiling problem for the gentoo folks where the compilers do actually complain if you try to use functions which the compiler does not (yet) know about.

Comment (RincewindsHat) [reply plugins/sslutils.c:194]: But for more clearness, I can separate this

Comment (sni) [reply plugins/sslutils.c:194]: nope, it's fine. Just wanted to know

**Review by sni — APPROVED** [review]
looks fine then :)
